engineapi/proxy.go

```diff
@@ -19,77 +23,296 @@ import (
+					if err != nil {
+						cLog.WithError(err).Debug("Cannot replace proxy client")
+					} else {
+						cLog.Debug("Replacing with new proxy client")
+						h.NewHandleClient(im, newIndex)
+					}
+
+					client.SetNextResync(h.clock, h.resyncPeriod)
+					break
+				}
+
+				h.gc(instanceManagerImage, clientIndex)
+
+				if client.ShouldCreate(im) {
+					cLog.Debug("Creating proxy client")
+
+					if err := client.Start(clientIndex, im, h.logger, h.ds, h.stopCh); err != nil {
+						cLog.Warn("Failed to start proxy client")
+					}
+				}
+
+				client.SetNextResync(h.clock, h.resyncPeriod)
+			}
+
+			exceedMaxConnection := true
+			for _, client := range handle.Clients {
+				if client.GetInUse() >= MaxProxyClientConcurrentInUse {
+					continue
+				}
+
+				exceedMaxConnection = false
+				break
+			}
+			if len(handle.Clients) != 0 && exceedMaxConnection {
+				log.Debugf("Exceed proxy client max concurrent use limit:%v", MaxProxyClientConcurrentInUse)
+
+				newIndex, err := h.getNewClientIndex(handle.Clients)
+				if err != nil {
+					log.WithError(err).Debugf("Cannot add more proxy client")
+				} else {
+					log.Debugf("Add proxy client-%v to handle concurrent use", newIndex)
+					h.NewHandleClient(im, newIndex)
+				}
+			}
+		}
+	}
+}
+
+func (h *EngineClientProxyHandler) gc(instanceManagerImage string, clientIndex int) {
+	client, exist := h.GetHandleClient(instanceManagerImage, clientIndex)
+	if !exist {
+		h.logger.Debugf("Proxy %v client-%v already gone", instanceManagerImage, clientIndex)
+		return
+	}
+
+	log := client.GetLogger()
+
+	nextGC := client.GetNextGC()
+	now := h.clock.Now()
+	if nextGC == nil || now.Before(*nextGC) {
+		return
+	}
+
+	msg := ""
+	if client.GetInUse() != 0 {
+		msg = fmt.Sprintf("Cannot stop proxy client with %v still in use", client.GetInUse())
+		if client.GetGCRetry() < ProxyGCMaxRetry {
+			log.Warnf(msg)
+			client.SetNextGC(h.clock)
+			return
+		}
+	}
+
+	log.Debug("Stopping proxy client")
+	if err := client.Stop(); err != nil {
+		msg = "Failed to stop proxy client"
+		if client.GetGCRetry() < ProxyGCMaxRetry {
+			log.WithError(err).Warnf(msg)
+			client.SetNextGC(h.clock)
+			return
+		}
+	}
+
+	if msg != "" {
+		log.Warn("Exceed max retry limit, force delete proxy client")
+	}
+
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	delete(h.handle[instanceManagerImage].Clients, clientIndex)
+	log.Debug("Deleted proxy client")
+}
+
+func (h *EngineClientProxyHandler) getNewClientIndex(clients map[int]EngineClientProxy) (int, error) {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	if len(clients) >= MaxProxyClientPerImage {
+		return 0, errors.Errorf("exceed max proxy client per image limit:%v", MaxProxyClientPerImage)
+	}
+
+	maxIndex := 0
+	for index := range clients {
+		if index <= maxIndex {
+			continue
+		}
+
+		maxIndex = index
+	}
+	return maxIndex + 1, nil
+}
+
+func (h *EngineClientProxyHandler) GetHandledInstanceManagerImages() map[string]EngineClientProxys {
```

Comment:
The naming should be aligned with it's returned value. ex: `GetEngineClientProxies` instead. The same as other method names.

Reply:
Updated.